lib/asyncSender.ts

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache License. See License in the project root for license information.
+
+import {
+   Delivery, Message, Sender as RheaSender, SessionEvents
+} from "rhea";
+import * as log from "./log";
+import { BaseSender, BaseSenderOptions } from "./sender";
+import { SenderEvents } from "rhea";
+import { OnAmqpEvent, EventContext } from "./eventContext";
+import { Session } from "./session";
+import { OperationTimeoutError } from "./operationTimeoutError";
+
+/**
+ * @internal
+ */
+export interface PromiseLike {
+  resolve: (value?: any) => void;
+  reject: (reason?: any) => void;
+  timer: NodeJS.Timer;
+}
+
+/**
+ * Describes the event listeners that can be added to the AsyncSender.
+ * @interface Sender
+ */
+export declare interface AsyncSender {
+  on(event: SenderEvents, listener: OnAmqpEvent): this;
+}
+
+export interface AsyncSenderOptions extends BaseSenderOptions {
+  /**
+   * The duration in which the promise to send the message should complete (resolve/reject).
+   * If it is not completed, then the Promise will be rejected after timeout occurs.
+   * Default: `20 seconds`.
+   */
+  messageTimeoutInSeconds?: number;
+}
+
+/**
+ * Describes the async version of the sender where one can await on the message being sent.
+ * @class AsyncSender
+ */
+export class AsyncSender extends BaseSender {
```

Comment:
Did we consider adding another function say `sendAsync()` to the existing `Sender` instead of creating a new Sender type?

Thinking from a user's perspective, it feels like an overkill to have to create a whole new type of sender for a different behavior.

Reply:
Yup, I had designed it like that initially. However, there were some disadvantages to that design.
- To be able to resolve/reject a promise the sender needs to (internally) add handlers for `"accepted"`, `"rejected"`, `"modified"`, `"released"`. In the current `Sender`, the user can also provide handlers for the above events. If both are added then the event is handled at both the places and this will create confusion.
- The above point, could have been handled by creating an option like `asyncmode: true | false` and then throw appropriate error message from `Sender.send()` and `Sender.sendAsync()` if the user is trying to use the method which was not supposed to be used. We would also have to throw an error if `asyncmode: true` and the user provided handlers for `"accepted"`, `"rejected"`, `"modified"`, `"released"`.
- I feel this makes things more complex as of today and will only make it worse if more options need to be added in the future.

Thus having two senders:
- `Sender` - bare bones sender, where the user wants to manage everything (handling disconnects, ensure successful delivery of message, retry on send operation) end to end.
- `AsyncSender` - where the sender manages some of the above mentioned things for the user.
makes it very clear for the user, w.r.t why they want to use one over another.

Reply:
As discussed, will be updating `AsyncSender` to `AwaitableSender`